File: web/src/api/activities.ts
import api from "./client";
import type { Card } from "@/types/card";

export interface PresenceStatus {
    /** false tant que la chance de présence n'est pas débloquée (niveau). */
    unlocked?: boolean;
    multiplier: number;
    max_multiplier: number;
    full_after_hours: number;
    present_seconds: number;
    chest: { coins: number; dust: number; hours: number; cap_hours: number };
}

export interface ExpeditionEstimate {
    coins: number;
    dust: number;
    booster_chance: number;
    rare_card_chance: number;
    power_factor: number;
}

export interface ExpeditionOut {
    id: number;
    slot: number;
    duration_minutes: number;
    ends_at: string;
    remaining_seconds: number;
    done: boolean;
    total_power: number;
    cards: Card[];
    estimate: ExpeditionEstimate;
}

export interface ExpeditionsOverview {
    slots: { slot: number; expedition: ExpeditionOut | null }[];
    durations: number[];
    max_cards: number;
    locked_card_ids: string[];
}

export interface ExpeditionReward {
    coins: number;
    dust: number;
    booster_id: string | null;
    booster_name: string | null;
    card: Card | null;
}

export interface WorkshopState {
    taps: number;
    taps_per_gauge: number;
    gauges_today: number;
    gauges_per_day: number;
    fragments: number;
    fragments_per_booster: number;
    coins_per_gauge: number;
    accepted?: number;
    reward: { coins: number; dust: number; gauges: number; boosters: number };
    booster_id?: string | null;
    booster_name?: string | null;
}

export interface HigherLowerGameState {
    id: number;
    status: "active" | "cashed" | "lost";
    resource_id: string;
    stake: number;
    step: number;
    max_steps: number;
    min_cashout_step: number;
    total_multiplier: number;
    current_card: Card;
    cashout_value: number;
    can_cashout: boolean;
    /** Gain d'une bonne réponse (null = réponse impossible). */
    odds: { higher?: number | null; lower?: number | null };
    outcome?: "win" | "tie" | "lose";
    won_multiplier?: number | null;
    previous_card?: Card;
}

export interface HigherLowerState {
    game: HigherLowerGameState | null;
    min_stake: number;
    max_stake: number;
    max_steps: number;
    min_cashout_step: number;
}

export interface WheelState {
    free_available: boolean;
    extra_spins_used: number;
    extra_spins_per_day: number;
    extra_spin_cost: number;
    segments: { label: string; kind: "resource" | "booster" | "reroll" }[];
}

export interface WheelSpin {
    index: number;
    reward: {
        kind: "resource" | "booster" | "reroll";
        label: string;
        amount: number;
        resource_id: string | null;
        booster_id: string | null;
        booster_name: string | null;
    };
    state: WheelState;
}

export interface MachineResourceNeed {
    resource_id: string;
    name: string;
    amount: number;
}

export interface MachineUpgrade {
    kind: string;
    label: string;
    next: string;
    level: number;
    cost: number;
    chance: number;
    /** Ressource obligatoire à ce cran (null avant le cran de départ). */
    required: MachineResourceNeed | null;
    /** Ressources qu'on peut ajouter pour augmenter la réussite (par unité). */
    bonus_options: { resource_id: string; name: string; per_unit: number }[];
    /** Réussite maximale atteignable en ajoutant des ressources, à ce cran. */
    cap: number;
}

export interface MachineItem {
    item: "booster" | "reroll";
    booster_id?: string;
    bonus_id?: number | null;
    token_id?: number;
    name: string;
    detail: string;
    quantity: number;
    upgrades: MachineUpgrade[];
}

export interface MachineState {
    day: number;
    length: number;
    today: { kind: string; label: string }[];
    event: { id: string; label: string } | null;
    cycle_upgrades: string[];
    cycle_events: string[];
    items: MachineItem[];
}

export interface MachineResult {
    success: boolean;
    destroyed: boolean;
    cost: number;
    chance: number;
    result: string | null;
    spent: Record<string, number>;
    state: MachineState;
}

export interface ConverterState {
    daily_uses: number;
    uses_left: number;
    pairs: { from: string; to: string; give: number; get: number; max_in: number }[];
}

export const activitiesApi = {
    machine: () => api.get<MachineState>("/activities/machine").then((r) => r.data),
    upgrade: (item: MachineItem, kind: string, extra: Record<string, number> = {}) =>
        api.post<MachineResult>("/activities/machine/upgrade", {
            item: item.item, kind, booster_id: item.booster_id, bonus_id: item.bonus_id ?? null, token_id: item.token_id, extra,
        }).then((r) => r.data),
    converter: () => api.get<ConverterState>("/activities/converter").then((r) => r.data),
    convert: (fromId: string, toId: string, amount: number) =>
        api.post<{ spent: number; gained: number; state: ConverterState }>("/activities/converter", {
            from_id: fromId, to_id: toId, amount,
        }).then((r) => r.data),
    presence: () => api.get<PresenceStatus>("/activities/presence").then((r) => r.data),
    ping: () => api.post<PresenceStatus>("/activities/presence/ping").then((r) => r.data),
    expeditions: () => api.get<ExpeditionsOverview>("/activities/expeditions").then((r) => r.data),
    startExpedition: (slot: number, durationMinutes: number, cardIds: string[]) =>
        api.post<ExpeditionOut>("/activities/expeditions", {
            slot, duration_minutes: durationMinutes, card_ids: cardIds,
        }).then((r) => r.data),
    claimExpedition: (id: number) =>
        api.post<ExpeditionReward>(`/activities/expeditions/${id}/claim`).then((r) => r.data),
    workshop: () => api.get<WorkshopState>("/activities/workshop")
        .then((r) => ({ ...r.data, reward: { coins: 0, dust: 0, gauges: 0, boosters: 0 } })),
    workshopTaps: (count: number) =>
        api.post<WorkshopState>("/activities/workshop/taps", { count }).then((r) => r.data),
    higherLower: () => api.get<HigherLowerState>("/activities/higher-lower").then((r) => r.data),
    startHigherLower: (resourceId: string, stake: number) =>
        api.post<HigherLowerGameState>("/activities/higher-lower", { resource_id: resourceId, stake }).then((r) => r.data),
    guessHigherLower: (id: number, guess: "higher" | "lower") =>
        api.post<HigherLowerGameState>(`/activities/higher-lower/${id}/guess`, { guess }).then((r) => r.data),
    cashoutHigherLower: (id: number) =>
        api.post<HigherLowerGameState>(`/activities/higher-lower/${id}/cashout`).then((r) => r.data),
    wheel: () => api.get<WheelState>("/activities/wheel").then((r) => r.data),
    spinWheel: () => api.post<WheelSpin>("/activities/wheel/spin").then((r) => r.data),
    claimChest: () =>
        api.post<{ coins: number; dust: number; hours: number }>("/activities/presence/chest").then((r) => r.data),
};

File: backend/app/api/activities.py
"""
Routes — activités : présence (bonus de chance, coffre d'absence).
Les autres activités (expéditions, atelier, mini-jeux) s'ajoutent ici.
"""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user, require_admin
from app.core.ratelimit import rate_limit
from app.database import get_session
from app.models.user import User
from pydantic import BaseModel, Field

from app.services import activities_config, expeditions, machine, minigames, presence_bonus, workshop
from app.services import unlocks

router = APIRouter()
admin_router = APIRouter(dependencies=[Depends(require_admin)])


@router.get("/presence")
async def presence_status(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await presence_bonus.status(session, user)


@router.post("/presence/ping", dependencies=[Depends(rate_limit(10, 60))])
async def presence_ping(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    """Signal envoyé toutes les 30 s par l'appli tant qu'elle est affichée."""
    return await presence_bonus.ping(session, user)


@router.post("/presence/chest", dependencies=[Depends(rate_limit(10, 60))])
async def claim_absence_chest(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    await unlocks.require(session, user, "absence_chest")
    return await presence_bonus.claim_chest(session, user)


class ExpeditionStartBody(BaseModel):
    slot: int = Field(ge=0)
    duration_minutes: int = Field(ge=1)
    card_ids: list[str] = Field(min_length=1, max_length=10)


@router.get("/expeditions")
async def expeditions_overview(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await expeditions.overview(session, user)


@router.post("/expeditions", dependencies=[Depends(rate_limit(20, 60))])
async def start_expedition(
    body: ExpeditionStartBody,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    await unlocks.require(session, user, "expeditions")
    return await expeditions.start(session, user, body.slot, body.duration_minutes, body.card_ids)


@router.post("/expeditions/{expedition_id}/claim", dependencies=[Depends(rate_limit(20, 60))])
async def claim_expedition(
    expedition_id: int,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await expeditions.claim(session, user, expedition_id)


class TapBody(BaseModel):
    count: int = Field(ge=0, le=1000)


@router.get("/workshop")
async def workshop_status(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await workshop.status(session, user)


@router.post("/workshop/taps", dependencies=[Depends(rate_limit(120, 60))])
async def workshop_taps(
    body: TapBody,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    """Lot de taps envoyé environ une fois par seconde par l'appli."""
    await unlocks.require(session, user, "workshop")
    return await workshop.tap(session, user, body.count)


class HigherLowerStartBody(BaseModel):
    resource_id: str
    stake: int = Field(ge=1)


class HigherLowerGuessBody(BaseModel):
    guess: str = Field(pattern="^(higher|lower)$")


@router.get("/higher-lower")
async def higher_lower_state(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await minigames.higher_lower_state(session, user)


@router.post("/higher-lower", dependencies=[Depends(rate_limit(30, 60))])
async def higher_lower_start(
    body: HigherLowerStartBody,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    await unlocks.require(session, user, "higher_lower")
    return await minigames.higher_lower_start(session, user, body.resource_id, body.stake)


@router.post("/higher-lower/{game_id}/guess", dependencies=[Depends(rate_limit(60, 60))])
async def higher_lower_guess(
    game_id: int,
    body: HigherLowerGuessBody,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await minigames.higher_lower_guess(session, user, game_id, body.guess)


@router.post("/higher-lower/{game_id}/cashout", dependencies=[Depends(rate_limit(30, 60))])
async def higher_lower_cashout(
    game_id: int,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await minigames.higher_lower_cashout(session, user, game_id)


@router.get("/wheel")
async def wheel_state(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    return await minigames.wheel_state(session, user)


@router.post("/wheel/spin", dependencies=[Depends(rate_limit(20, 60))])
async def wheel_spin(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    await unlocks.require(session, user, "wheel")
    return await minigames.wheel_spin(session, user)


class UpgradeBody(BaseModel):
    item: str = Field(pattern="^(booster|reroll)$")
    kind: str = Field(max_length=40)
    booster_id: str | None = None
    bonus_id: int | None = None
    token_id: int | None = None


class ConvertBody(BaseModel):
    from_id: str = Field(max_length=30)
    to_id: str = Field(max_length=30)
    amount: int = Field(ge=1, le=10_000_000)


@router.get("/machine")
async def machine_state(user: User = Depends(get_current_user), session: AsyncSession = Depends(get_session)):
    return await machine.machine_state(session, user)


@router.post("/machine/upgrade", dependencies=[Depends(rate_limit(30, 60))])
async def machine_upgrade(body: UpgradeBody, user: User = Depends(get_current_user),
                          session: AsyncSession = Depends(get_session)):
    await unlocks.require(session, user, "machine")
    return await machine.upgrade(session, user, body.item, body.kind, body.booster_id, body.bonus_id, body.token_id)


@router.get("/converter")
async def converter_state(user: User = Depends(get_current_user), session: AsyncSession = Depends(get_session)):
    return await machine.converter_state(session, user)


@router.post("/converter", dependencies=[Depends(rate_limit(20, 60))])
async def convert(body: ConvertBody, user: User = Depends(get_current_user), session: AsyncSession = Depends(get_session)):
    await unlocks.require(session, user, "converter")
    return await machine.convert(session, user, body.from_id, body.to_id, body.amount)


@admin_router.get("/config")
async def get_activities_config(session: AsyncSession = Depends(get_session)):
    return await activities_config.get_config(session)


@admin_router.put("/config")
async def put_activities_config(values: dict, session: AsyncSession = Depends(get_session)):
    return await activities_config.save_config(session, values)
